source/tests/integration/ni_fake_non_ivi_service_tests_endtoend.cpp

```diff
@@ -0,0 +1,235 @@
+#include <gtest/gtest.h>
+#include <nifake_non_ivi/nifake_non_ivi_mock_library.h>
+#include <nifake_non_ivi/nifake_non_ivi_service.h>
+
+#include <atomic>
+#include <memory>
+#include <thread>
+
+namespace ni {
+namespace tests {
+namespace integration {
+
+using namespace nifake_non_ivi_grpc;
+
+// Test NiFakeNonIviService using a client stub.
+class NiFakeNonIviServiceTests_EndToEnd : public ::testing::Test {
+ public:
+  using FakeResourceRepository = nidevice_grpc::SessionResourceRepository<FakeHandle>;
+  nidevice_grpc::SessionRepository session_repository_;
+  std::shared_ptr<FakeResourceRepository> resource_repository_;
+  ni::tests::unit::NiFakeNonIviMockLibrary library_;
+  NiFakeNonIviService service_;
+  std::unique_ptr<::grpc::Server> server_;
+  std::unique_ptr<NiFakeNonIvi::Stub> stub_;
+  std::atomic<bool> shutdown_{false};
+
+  NiFakeNonIviServiceTests_EndToEnd()
+      : session_repository_(),
+        resource_repository_(std::make_shared<FakeResourceRepository>(&session_repository_)),
+        library_(),
+        service_(&library_, resource_repository_),
+        server_(start_server()),
+        stub_(NiFakeNonIvi::NewStub(server_->InProcessChannel(::grpc::ChannelArguments())))
+  {
+  }
+
+  virtual ~NiFakeNonIviServiceTests_EndToEnd()
+  {
+    shutdown();
+  }
+
+  void shutdown()
+  {
+    if (!shutdown_) {
+      shutdown_ = true;
+      server_->Shutdown();
+    }
+  }
+
+  std::unique_ptr<::grpc::Server> start_server()
+  {
+    ::grpc::ServerBuilder builder;
+    builder.RegisterService(&service_);
+    return builder.BuildAndStart();
+  }
+
+  std::unique_ptr<NiFakeNonIvi::Stub>& stub()
+  {
+    return stub_;
+  }
+};
+
+TEST_F(NiFakeNonIviServiceTests_EndToEnd, ReadStream_ReturnsExpectedSequence)
+{
+  ::grpc::ClientContext context;
+  const auto START = 0;
+  const auto STOP = 100;
+  ReadStreamRequest request;
+  request.set_start(START);
+  request.set_stop(STOP);
+  auto reader = stub()->ReadStream(&context, request);
+
+  ReadStreamResponse response;
+  for (auto i = START; i <= STOP; ++i) {
+    reader->Read(&response);
+    EXPECT_EQ(response.value(), i);
+  }
+}
+
+TEST_F(NiFakeNonIviServiceTests_EndToEnd, ReadPartialStream_ReturnsExpectedSequence)
+{
+  ::grpc::ClientContext context;
+  const auto START = 0;
+  const auto STOP = 100;
+  ReadStreamRequest request;
+  request.set_start(START);
+  request.set_stop(STOP);
+  auto reader = stub()->ReadStream(&context, request);
+
+  ReadStreamResponse response;
+  for (auto i = START; i <= STOP / 2; ++i) {
+    reader->Read(&response);
+    EXPECT_EQ(response.value(), i);
+  }
+}
+
+TEST_F(NiFakeNonIviServiceTests_EndToEnd, ReadTwoStreamsAlternating_ReturnsExpectedSequences)
```

Comment:
Instead of "Alternating", could we use "Consecutively" or something? I was confused when we didn't go back and forth between `reader` and `second_reader` like the "Interleaved" test below.